### PR TITLE
Clean up `MemoryBank`

### PR DESCRIPTION
**Pull request by tilk**
This PR fixes something which bothered me for a long time. I couldn't wrap my head around why @lekcyjna123's `MemoryBank` implementation is so weird and complex. Finally, I understood the problem: in the version of Amaranth we are using, the read ports are transparent by default (!), and @lekcyjna123 basically tried to make non-transparent reads from a transparent read port.

What was changed:
* The `safe_writes` argument was replaced with `transparent`, which controls the read port transparency.
* The weird use of `ArgumentsToResultsZipper` and the risky use of an internal transaction which should always be able to run was removed. An explicit overflow management was added instead. 
* The test was changed so that it no longer uses implementation details of `MemoryBank` and handles transparency more naturally.
* The test was changed so that it is able to check consecutive reads and consecutive writes.
* Integrated the pipelining check into the main test.

### CODE REVIEW

File: transactron/lib/storage.py [hunk 1]
```diff
@@ -49,10 +48,8 @@ def __init__(
         granularity: Optional[int]
             Granularity of write, forwarded to Amaranth. If `None` the whole structure is always saved at once.
             If not, the width of `data_layout` is split into `granularity` parts, which can be saved independently.
-        safe_writes: bool
-            Set to `False` if an optimisation can be done to increase throughput of writes. This will cause that
-            writes will be reordered with respect to reads eg. in sequence "read A, write A X", read can return
-            "X" even when write was called later. By default `True`, which disable optimisation.
+        transparent: bool
+            Read port transparency, false by default.
```
Comment (piotro888) [inline transactron/lib/storage.py:52]: Quickly explaining transparency in docstring would be helpful

**Review by piotro888 — APPROVED** [review]
Nice simplification!

**Review by lekcyjna123 — APPROVED** [review]

File: transactron/lib/storage.py [hunk 2]
```diff
@@ -80,60 +79,45 @@ def elaborate(self, platform) -> TModule:
...
-        with self._internal_read_resp_trans.body(m, request=read_output_valid):
+        overflow_valid = Signal()
+        overflow_data = Signal(self.width)
+
+        # The read request method can be called at most twice when not reading the response.
+        # The first result is stored in the overflow buffer, the second - in the read value buffer of the memory.
+        # If the responses are always read as they arrive, overflow is never written and no stalls occur.
+
+        with m.If(read_output_valid & ~overflow_valid & self.read_req.run & ~self.read_resp.run):
             m.d.sync += read_output_valid.eq(0)
```
[the remark this answers is not in the document]
Comment (tilk) [reply transactron/lib/storage.py:93]: Could you elaborate? I don't understand this comment.

File: test/transactron/lib/test_transaction_lib.py [hunk 3]
```diff
@@ -139,22 +138,21 @@ def test_pipelining(self):
...
 
 class TestMemoryBank(TestCaseWithSimulator):
-    test_conf = [(9, 3, 3, 3, 14), (16, 1, 1, 3, 15), (16, 1, 1, 1, 16), (12, 3, 1, 1, 17)]
+    test_conf = [(9, 3, 3, 3, 14), (16, 1, 1, 3, 15), (16, 1, 1, 1, 16), (12, 3, 1, 1, 17), (9, 0, 0, 0, 18)]
 
-    parametrized_input = [tc + sf for tc, sf in itertools.product(test_conf, [(True,), (False,)])]
-
-    @parameterized.expand(parametrized_input)
-    def test_mem(self, max_addr, writer_rand, reader_req_rand, reader_resp_rand, seed, safe_writes):
+    @parameterized.expand(tc + tr for tc in test_conf for tr in [(False,), (True,)])
```
[the remark this answers is not in the document]
Comment (tilk) [reply test/transactron/lib/test_transaction_lib.py]: We will need a refactor commit some time later. But sure, I can change this now, the line is modified here anyway.

File: test/transactron/lib/test_transaction_lib.py [hunk 4]
```diff
@@ -163,83 +161,40 @@ def writer():
                 d = random.randrange(2**data_width)
                 a = random.randrange(max_addr)
                 yield from m.write.call(data=d, addr=a)
-                for _ in range(2):
+                for _ in range(2 if not transparent else 0):
```
[the remark this answers is not in the document]
Comment (tilk) [reply test/transactron/lib/test_transaction_lib.py:165]: Works for now. When `Settle` goes away, we'll have to rethink stuff. :(

File: transactron/lib/storage.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lekcyjna123) [reply transactron/lib/storage.py:93]: In line 109 there is:
```py
m.d.sync += read_output_valid.eq(1)
```
And this line will be run always if `~overflow_valid & self.read_req.run`, so this condition is superset of condition from line 92 (`read_output_valid & ~overflow_valid & self.read_req.run & ~self.read_resp.run`).

So always when condition from line 92 is true, then the condition to run `read_req` body is also true. so the line 93 is always overwritten by line 109.

Or I am wrong?

File: test/transactron/lib/test_transaction_lib.py [hunk 5]
```diff
@@ -163,83 +161,40 @@ def writer():
...
+                yield from self.random_wait(reader_req_rand)
 
         def reader_resp():
             for cycle in range(test_count):
+                for _ in range(2):
+                    yield Settle()
                 while not read_req_queue:
-                    yield from self.random_wait(reader_resp_rand, min_cycle_cnt=1)
+                    yield from self.random_wait(reader_resp_rand or 1, min_cycle_cnt=1)
+                    for _ in range(2):
```
[the remark this answers is not in the document]
Comment (tilk) [reply test/transactron/lib/test_transaction_lib.py]: Changed, to be safe.

File: transactron/lib/storage.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tilk) [reply transactron/lib/storage.py]: I read the code again and noticed the issue. Fixed.

**Pull request merged**